decorators_generators/HW_2.py

```diff
@@ -0,0 +1,455 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# # Задание 1 (2 балла)
+
+# Напишите класс `MyDict`, который будет полностью повторять поведение обычного словаря, за исключением того, что при итерации мы должны получать и ключи, и значения.
+# 
+# **Модули использовать нельзя**
+
+# In[2]:
+
+
+class DictIterator:
+    def __init__(self, dictionary):
+        self.dictionary = dictionary
+        self.__current_iteration = 0
+        self.length = len(dictionary)
+        
+    def __iter__(self):
+        return self
+        
+    def __next__(self):
+        if self.__current_iteration < self.length:
+            items_iter = iter(self.dictionary.items())
+            for i in range(self.__current_iteration):
+                next(items_iter)
+            result = next(items_iter)
+        else:
+            raise StopIteration
+        self.__current_iteration += 1
+        return result
+
+
+class MyDict(dict):
+    def __init__(self, dictionary):
+        super().__init__(dictionary)
+        self.dictionary = dictionary
+        
+    def __iter__(self):
+        return DictIterator(self.dictionary)
+
+
+# Или
+
+# In[3]:
+
+
+class MyDict(dict):        
+    def __iter__(self):
+        for items in self.items():
+            yield items
+
+
+# In[4]:
+
+
+dct = MyDict({"a": 1, "b": 2, "c": 3, "d": 25})
+for key, value in dct:
+    print(key, value)   
+
+
+# In[5]:
+
+
+for key, value in dct.items():
+    print(key, value)
+
+
+# In[6]:
+
+
+for key in dct.keys():
+    print(key)
+
+
+# In[7]:
+
+
+dct["c"] + dct["d"]
+
+
+# # Задание 2 (2 балла)
+
+# Напишите функцию `iter_append`, которая "добавляет" новый элемент в конец итератора, возвращая итератор, который включает изначальные элементы и новый элемент. Итерироваться по итератору внутри функции нельзя, то есть вот такая штука не принимается
+# ```python
+# def iter_append(iterator, item):
+#     lst = list(iterator) + [item]
+#     return iter(lst)
+# ```
+# 
+# **Модули использовать нельзя**
+
+# In[8]:
+
+
+def iter_append(iterator, item):
+    while True:
+        try:
+            yield next(iterator)
+        except StopIteration:
```

Comment:
Можно было так:
```suggestion
        yield from iterator
        yield item
```
По смыслу то же самое.